src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptionsUpdateHandler.cs

```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Reflection.Metadata;
+using System.Text.Json;
+
+[assembly: MetadataUpdateHandler(typeof(JsonSerializerOptionsUpdateHandler))]
+
+namespace System.Text.Json
+{
+    /// <summary>Handler used to clear JsonSerializerOptions reflection cache upon a metadata update.</summary>
+    internal static class JsonSerializerOptionsUpdateHandler
+    {
+        public static void BeforeUpdate(Type? type)
```

Comment:
Is the mechanism to invoke this method already implemented? Why's it called `BeforeUpdate`? Seems to me that it will be triggered _after_ a type has been updated.

Reply:
> Is the mechanism to invoke this method already implemented?

Designed but not implemented quite yet.  Should be soon.

> Why's it called BeforeUpdate? Seems to me that it will be triggered after a type has been updated.

The C# compiler produces the diff, which gets forwarded to the hot reload agent.  That agent, running in the target process, finds all the handlers, calls BeforeUpdate on each if BeforeUpdate exists, then calls ApplyUpdate with the diff to actually apply it, and then calls AfterUpdate on each if AfterUpdate exists.  Generally libraries like this will do whatever clearing they need in BeforeUpdate, and then higher level frameworks will regenerate whatever they need in AfterUpdate.

Reply:
If we want to be robust, I believe we need transactional-based metadata clearing since type metadata may be interdependent.

This could be done by flowing the concurrent dictionary instance throughout the graph when deserializing an object graph which then allows the actual global\static concurrent dictionary to be replaced (not cleared) that will cause  subsequent (de)serializations to re-build that cache in a consistent manner.

Reply:
It's already not perfect in general,  e.g. at the runtime level,  code could be using reflection concurrently with the type being updated, and race conditions will about.  The goal is "good enough" :-)